experimental/normalizer/src/lib.rs

```diff
@@ -464,30 +618,156 @@ where
+            return Err(NormalizerError::FutureExtension);
+        }
+
+        let ccc: DataPayload<CanonicalCombiningClassV1Marker> =
+            icu_properties::maps::get_canonical_combining_class(data_provider)?;
+
+        Ok(DecomposingNormalizer {
+            decompositions: PayloadHolder::Canonical(decompositions),
+            tables,
+            supplementary_tables: None,
+            ccc,
+        })
+    }
+
+    /// NFKD constructor.
+    pub fn try_new_nfkd<D>(data_provider: &D) -> Result<Self, NormalizerError>
+    where
+        D: ResourceProvider<CompatibilityDecompositionDataV1Marker>
+            + ResourceProvider<CanonicalDecompositionTablesV1Marker>
+            + ResourceProvider<CompatibilityDecompositionTablesV1Marker>
+            + ResourceProvider<icu_properties::provider::CanonicalCombiningClassV1Marker>
+            + ?Sized,
+    {
+        let decompositions: DataPayload<CompatibilityDecompositionDataV1Marker> = data_provider
+            .load_resource(&DataRequest::default())?
+            .take_payload()?;
+        let tables: DataPayload<CanonicalDecompositionTablesV1Marker> = data_provider
+            .load_resource(&DataRequest::default())?
+            .take_payload()?;
+        let supplementary_tables: DataPayload<CompatibilityDecompositionTablesV1Marker> =
+            data_provider
+                .load_resource(&DataRequest::default())?
+                .take_payload()?;
+
+        if tables.get().scalars16.len()
+            + tables.get().scalars32.len()
+            + supplementary_tables.get().scalars16.len()
+            + supplementary_tables.get().scalars32.len()
+            > 0xFFF
+        {
+            // The data is from a future where there exists a normalization flavor whose
+            // complex decompositions take more than 0xFFF but fewer than 0x1FFF code points
+            // of space. If a good use case from such a decomposition flavor arises, we can
+            // dynamically change the bit masks so that the length mask becomes 0x1FFF instead
+            // of 0xFFF and the all-non-starters mask becomes 0 instead of 0x1000. However,
+            // since for now the masks are hard-coded, error out.
+            return Err(NormalizerError::FutureExtension);
+        }
+
+        let ccc: DataPayload<CanonicalCombiningClassV1Marker> =
+            icu_properties::maps::get_canonical_combining_class(data_provider)?;
+
+        Ok(DecomposingNormalizer {
+            decompositions: PayloadHolder::Compatibility(decompositions),
+            tables,
+            supplementary_tables: Some(supplementary_tables),
+            ccc,
+        })
+    }
+
+    /// NFKD_CaseFold constructor.
+    ///
+    /// This is a special building block normalization for IDNA. It is the decomposed counterpart of
+    /// the NFKC_CaseFold normalization specified for IDNA except that default ignorables map to themselves
+    /// instead of mapping to the empty string.
```

Comment:
I don't think this is a good idea.

First: If you name it like the standard thing but it doesn't behave like the standard thing (because it doesn't map default ignorables away), then it's grossly misleading.

Second: It is misleading to suggest that this is how you implement IDNA. Either you implement vanilla IDNA2008, and don't have a mapping table. Or you implement `UTS #46` which does have a mapping table, but it's not the same as the one used for NFKC_CaseFold.

In ICU, the `UTS #46` implementation does use a normalization data file, one derived from the Unicode IdnaMappingTable.txt, and that does map default ignorables to nothing. There is nothing in the Unicode normalization algorithm forbidding a mapping to nothing/empty.
- https://github.com/unicode-org/icu/blob/main/icu4c/source/data/unidata/norm2/uts46.txt
- https://github.com/unicode-org/icu/blob/main/icu4c/source/common/uts46.cpp

Reply:
> I don't think this is a good idea.

It may well be a bad idea. I wanted to pursue this half-thought idea now to see what the normalizer that I have written with NFD focus can do for non-NFD use cases with different data.

So far I have learned:
1. The original data bit allocation precluded NFKD data, so this PR fixes that.
2. This normalizer implementation can't deal with mapping a character to emptiness.

> First: If you name it like the standard thing but it doesn't behave like the standard thing (because it doesn't map default ignorables away), then it's grossly misleading.

I am uncomfortable with the name, but I was unimaginative.

> Second: It is misleading to suggest that this is how you implement IDNA. Either you implement vanilla IDNA2008, and don't have a mapping table. Or you implement `UTS #46` which does have a mapping table, but it's not the same as the one used for NFKC_CaseFold.
> 
> In ICU, the `UTS #46` implementation does use a normalization data file, one derived from the Unicode IdnaMappingTable.txt, and that does map default ignorables to nothing. There is nothing in the Unicode normalization algorithm forbidding a mapping to nothing/empty.
> 
>     * https://github.com/unicode-org/icu/blob/main/icu4c/source/data/unidata/norm2/uts46.txt
> 
>     * https://github.com/unicode-org/icu/blob/main/icu4c/source/common/uts46.cpp

This intends to be data extracted from ICU's UTS 46 normalizer, but I may have extracted the wrong data. I extracted this data from `Normalizer2::getNFKCCasefoldInstance()`, which did normalize default ignorables to the empty string.

(The inability to deal with mapping to the empty string is a constraint of this normalizer implementation. Not dealing with mappings to empty simplifies things and it makes sense to keep things simple until Unicode introduces mappings to empty in NFD or NFKD, and hopefully that's never.)

The idea is that when IDNA has the mapping step followed by NFC, the upcoming composing normalizer with this data could be used for a combination of a part of the "mapped" part of the "Map" step and the following NFC. Since the Map step before normalization has to deal with "disallowed" and "ignored" anyway, it has to inspect each character before normalization anyway, so it seems OK to handle the ignoring of default ignorables in that step.

Reply:
I made the following changes:
* Extracted data on the icuexportdata side from the right ICU4C normalization.
* Renamed to indicate UTS 46 specifically.
* Made the normalizer constructor test-only, since (apart from testing), this data should only be used for UTS 46, which needs also the canonical composition step.
* Documented the differences from ICU4C's UTS 46 normalization to say what processing is the responsibility of the previous processing step in an UTS 46 implementation that uses this data.
* Removed the documentation suggestion that this might be useful for anything other than UTS 46.